### PR TITLE
Bratseth/lazy summary decoding

### PR DESCRIPTION
**Pull request by bratseth**
@arnej27959 please review

This keeps the summary data as a slime structure in FastHit. The API surface of Hit regarding fields is pretty large. Most of the code (in FastHit) is just to preserve it without unnecessary decoding.

### CODE REVIEW

File: container-search/src/main/java/com/yahoo/prelude/fastsearch/FastHit.java [hunk 1]
```diff
@@ -205,4 +354,402 @@ public static String asHexString(GlobalId gid) {
...
+                for (int i = 0; i < index; i++) {
+                    if (hit.summaries.get(i).type.fieldNames().contains(fieldName))
+                        return true;
+                }
+                return false;
+            }
+
+        }
+
+        private static class SummaryFieldDataIterator extends SummaryDataIterator<Map.Entry<String, Object>> {
```
Comment (arnej27959) [inline container-search/src/main/java/com/yahoo/prelude/fastsearch/FastHit.java]: A short comment on the purpose of the class would be good, maybe something like:
Iterate over the fields of a single summary data object.
Also consider changing name to 'SummaryDataFieldIterator'.

File: container-search/src/main/java/com/yahoo/prelude/fastsearch/FastHit.java [hunk 2]
```diff
@@ -205,4 +354,402 @@ public static String asHexString(GlobalId gid) {
...
+
+                @Override
+                public Object setValue(Object value) { throw new UnsupportedOperationException(); }
+
+            }
+
+        }
+
+        /** A wrapper of a field iterator which converts to the expected field types. Read only. */
+        private static class SummaryFieldNameDataIterator extends SummaryDataIterator<String> {
```
Comment (arnej27959) [inline container-search/src/main/java/com/yahoo/prelude/fastsearch/FastHit.java]: A short comment on the purpose of the class would be good, maybe something like:
Iterate over the field names of a single summary data object.
Also consider changing name to 'SummaryDataFieldNameIterator'.

File: container-search/src/main/java/com/yahoo/prelude/fastsearch/FastHit.java [hunk 3]
```diff
@@ -205,4 +354,402 @@ public static String asHexString(GlobalId gid) {
...
+                currentIterator = iteratorFor(hit.summaries.get(currentSummaryDataIndex));
+                if (currentIterator.hasNext())
+                    return true;
+            }
+            return false;
+        }
+
+    }
+
+    private static class FieldIterator extends SummaryIterator<Map.Entry<String, Object>> {
```
Comment (arnej27959) [inline container-search/src/main/java/com/yahoo/prelude/fastsearch/FastHit.java]: Again a bit of comment would be good, something like:
Iterates over all fields in the hit (including all attached SummaryData) using sub-iterators

File: container-search/src/main/java/com/yahoo/prelude/fastsearch/FastHit.java [hunk 4]
```diff
@@ -205,4 +354,402 @@ public static String asHexString(GlobalId gid) {
...
+
+        @Override
+        protected Iterator<Map.Entry<String, Object>> iteratorFor(SummaryData summary) {
+            return summary.fieldIterator();
+        }
+
+    }
+
+
+    private static class FieldNameIterator extends SummaryIterator<String> {
```
Comment (arnej27959) [inline container-search/src/main/java/com/yahoo/prelude/fastsearch/FastHit.java]: Iterates over all field names in the hit (including all attached SummaryData) using sub-iterators

**Review by arnej27959 — APPROVED** [review]
overall looks nice, just minor nits (would like some comments as indicated).  it's hard to see if the logic for resolving which fields are visible from where is correct, especially with the removedFields override,
so I hope your unit tests have really good coverage :-)

**Comment by bratseth** [comment]
Thanks for the name suggestions - better. I did that and added more javadoc.

**Pull request merged**